SITL: add copter-tailsitter

IamPete1 opened the pull request:
This adds a copter tailsitter model to SITL. Moving towards adding some tailsitters to autotest. Seems to fly fine and completes the vtol mission, have run Qautotune but parameters and tune could be improved.

kd0aij commented:
Regarding autotune: how accurate is the simulation? I couldn't find any moments of inertia in the code.
Also, what log records should I look at to judge the success of autotuning? Is there something that shows the commanded rates/angles or the rate/angle errors?

IamPete1 commented on libraries/SITL/SIM_QuadPlane.cpp:
```diff
@@ -67,6 +69,12 @@ QuadPlane::QuadPlane(const char *home_str, const char *frame_str) :
         // fwd motor gives zero thrust
         thrust_scale = 0;
+    } else if (strstr(frame_str, "-copter_tailsitter")) {
+        frame_type = "+";
```
yeah, i think so

IamPete1 commented:
I think its not accurate enough to bother getting a perfect tune, once you exit you should get a normal log just as you would IRL, you can looks at the rates ect then. Its a bit tricky to judge in real time tho.

IamPete1 commented:
This is now two commits and inherits plane default params. Only odd thing is I have to turn off arming checks due to the EKF failing. I think this is due to the euler angles and the tailsitter ground state

WickedShell commented on Tools/autotest/default_params/quadplane-copter_tailsitter.parm:
```diff
@@ -0,0 +1,36 @@
+Q_A_RAT_YAW_IMAX 0.500000
+Q_A_RAT_YAW_P    0.391919
+Q_TAILSIT_THSCMX 1.5
+ARMING_CHECK     0
```
Arming check 0? Seems... questionable.

IamPete1 commented:
I have fixed the spinning round on the ground by only rotating pitch to 85 deg rather than 90. Then the yaw no longer needs to be zeroed. The issue was trying to rotate by >=90 deg using euler angles, the other two axis can no longer be solved (i think). This still gets EKF angle estimate error on arming, I have changed the arming check to only ignore this. Again i think this is due to the difference between ardupilots quaternions and SILT's euler angles (probably but not sure)

kd0aij commented:
I think I have a much simpler fix for the ground behavior now with commit 92aec5b34b58c4b56dc33ff702658408b20d4ed3
This is a plot of attitude with log_disarmed enabled. There appears to be a problem with yaw in flight though; and I don't think it was caused by the ground behavior.
![tailsitter-sitl](https://user-images.githubusercontent.com/2300221/56088504-86430300-5e3f-11e9-9333-c43b24d10598.png)

kd0aij commented:
related PR: https://github.com/ArduPilot/ardupilot/pull/11113 to address error:

```
>>>> FAILED STEP: fly.QuadPlane at Sun Apr 14 09:49:55 2019 ('list' object has no attribute 'startswith')
Traceback (most recent call last):
  File "./Tools/autotest/autotest.py", line 589, in run_tests
    if run_step(step):
  File "./Tools/autotest/autotest.py", line 386, in run_step
    return tester_class_map[step](binary, **fly_opts).autotest()
  File "/home/markw/linux_git/kd0aij/ardupilot/Tools/autotest/common.py", line 2672, in autotest
    ret = self.run_tests(tests)
  File "/home/markw/linux_git/kd0aij/ardupilot/Tools/autotest/common.py", line 2561, in run_tests
    self.init()
  File "/home/markw/linux_git/kd0aij/ardupilot/Tools/autotest/common.py", line 1869, in init
    defaults_file=self.defaults_filepath(),
  File "/home/markw/linux_git/kd0aij/ardupilot/Tools/autotest/quadplane.py", line 44, in defaults_filepath
    return os.path.join(testdir, defaults_file)
  File "/home/markw/anaconda2/lib/python2.7/posixpath.py", line 68, in join
    if b.startswith('/'):
AttributeError: 'list' object has no attribute 'startswith'
check step:  fly.QuadPlane
```

davidbuzz commented:
tests not passing.

peterbarker commented:
This was cleanly rebasing until today.  Should we mark this DevCall to work out what to do with it?

IamPete1 commented:
re-based but not re-tested

IamPete1 commented:
does not fly, I have not taken the time to work out what is going on.
